R/toolISOhistorical.R

```diff
@@ -115,56 +116,253 @@ toolISOhistorical <- function(m, mapping = NULL, additional_mapping = NULL, over
+                                          c("fromISO", "toISO")],
+                             from = "fromISO", to = "toISO",
+                             dim = 1,
+                             weight = weight[setNames(list(c(getItems(z[setNames(list(c(tr$fromISO)), mainDimName), , ],
+                                                                      dim = secdDimName))), secdDimName), , ] +
+                               weight2[setNames(list(c(getItems(z[setNames(list(c(tr$fromISO)), mainDimName), , ],
+                                                                   dim = secdDimName))), secdDimName), , ], # nolint #add small weight to disaggregate properly
+                             wdim = 1)
+
+        # aggregation of countries
+      } else {
+        zTr <- toolAggregate(z[setNames(list(c(tr$fromISO)), mainDimName), subTime, ],
+                             bilatMapping[is.element(bilatMapping$toISO,
+                                                     as.vector(outer(tr$toISO, getItems(z[setNames(list(c(tr$fromISO)), mainDimName), subTime, ], dim = secdDim), # nolint
+                                                                     paste, sep = "."))),
+                                          c("fromISO", "toISO")],
+                             weight = NULL)
       }
 
-      mTr <- toolAggregate(m[a$fromISO, subTime, ],
-                           mapping[is.element(mapping$toISO, a$toISO), c("fromISO", "toISO")], weight = weight,
-                           negative_weight = "allow")
-      ## aggregation of countries
-    } else {
-      mTr <- toolAggregate(m[a$fromISO, subTime, ],
-                           mapping[is.element(mapping$toISO, a$toISO), c("fromISO", "toISO")], weight = NULL)
+      return(zTr)
     }
-    # fill data
-    if (overwrite == TRUE) {
-      m[a$toISO, subTime, ] <- mTr
-    } else {
-      m[a$toISO, subTime, ][is.na(m[a$toISO, subTime, ])] <- mTr[is.na(m[a$toISO, subTime, ])]
+
+    # loop over all transitions in dim 1
+    dims <- getSets(m)[c(1, 2)]
+    mainDim <- 1.1
+
+    for (a in tr) {
+      mTr <-   .histDim(m, tr = a, mainDim = mainDim, dims = dims)
+      # fill data
+      if (overwrite == TRUE) {
+        # only overwrite the exact combinations for which weighting existed, rest remains zero
+        addM <- setdiff(getItems(mTr, dim = 1), getItems(m, dim = 1))
+        if (length(addM) > 0) {
+          addMissM <- new.magpie(cells_and_regions = addM,
+                                 years = getItems(m, dim = 2),
+                                 names = getItems(m, dim = 3),
+                                 fill = 0)
+          m <- mbind(m, addMissM)
+        }
+
+        m[getItems(mTr, dim = 1), getYears(mTr), ] <- mTr
+      } else {
+        m[getItems(mTr, dim = 1), getYears(mTr), ][is.na(m[getItems(mTr, dim = 1), getYears(mTr), ])] <- mTr[is.na(m[getItems(mTr, dim = 1), getYears(mTr), ])] # nolint
```

Comment:
I'd suggest to split it up like so:
```suggestion
        a <- m[getItems(mTr, dim = 1), getYears(mTr), ]
        a[is.na(a)] <- mTr[is.na(a)]
```
The variable `a` should be renamed to something helpful though.